### PR TITLE
Add density metric from densMAP

### PR DESCRIPTION
**Pull request by michalk8**
### Submission type

* [ ] This submission adds a new dataset
* [ ] This submission adds a new method
* [x] This submission adds a new metric
* [ ] This submission adds a new task
* [ ] This submission adds a new Docker image
* [ ] This submission fixes a bug (link to related issue: )
* [ ] This submission adds a new feature not listed above

### Testing

* [x] This submission was written on a forked copy of SingleCellOpenProblems
* [ ] GitHub Actions "Run Benchmark" tests are passing on this base branch of this pull request (include link to passed test: )
* [x] If this pull request is not ready for review (including passing the "Run Benchmark" tests), I will open this PR as a draft (click on the down arrow next to the "Create Pull Request" button)

### Submission guidelines

* [x] This submission follows the guidelines in our [Contributing](../blob/master/CONTRIBUTING.md) document
* [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change

I did a quick search in the code+paper, the metric is correlation between log radius in the original space vs. log radius in the embedding space.
The code used for the radius in the orig. space is [here](https://github.com/hhcho/densvis/blob/3a68031658db276121638cd398919fdf7fea0914/densmap/densmap/densmap_.py#L1215) and in the emb. space is [here](https://github.com/hhcho/densvis/blob/3a68031658db276121638cd398919fdf7fea0914/densmap/densmap/densmap_.py#L1942) - principially it's the same.

What I propose is to extract the code from the 2nd link (pretty straight-forward). Then it's just running the extracted function
on the high dim. space, low dim. space + correlation coeff.

We should also include a regression test using densMAP and comparing the extracted function's Re (emb. radii) and Ro (original radii) to the ones from densMAP.

related issue: https://github.com/singlecellopenproblems/SingleCellOpenProblems/issues/291

### CODE REVIEW

**Comment by lazappi** [comment]
Nice 🎉 ! You have probably already looked but there should be an implementation in **densVis** https://github.com/hhcho/densvis and **umap** https://github.com/lmcinnes/umap. I doubt they are exposed for other people to use but should be a starting point at least.

**Comment by michalk8** [comment]
> Nice 🎉 ! You have probably already looked but there should be an implementation in densVis https://github.com/hhcho/densvis and umap https://github.com/lmcinnes/umap. I doubt they are exposed for other people to use but should be a starting point at least.

I've taken a closer look, but the 1st time I didn't find any, but maybe I'm still mistaken.

In any case, I've extracted the function + added a test to make sure it's the same, though not sure why calling `_calculate_radii`
was giving slightly different results (in terms of the correlation, if was off by 0.02).
Maybe for simplicity, I'd go back to using `_calculate_radii` rather than calling:
```
_, ro, _ = UMAP(
    n_neighbors=_K, random_state=_SEED, densmap=True, output_dens=True
).fit_transform(high_dim)
```
and test if the abs. delta of expected/actual correlation is within `.1` or similar.

**Comment by lazappi** [comment]
@michalk8 This looks good to me, can you please just fix the conflicts with the updates from the main repo? I think it's just another metric editing the same lines so should be fairly easy, or maybe just rebase off the current `dimred-metrics` branch here (whatever works).

Only other comment is that I think you borrowed code from the `umap` repo which has a BSD license. Might need to copy that into this file just to be safe.
